File: front-end/src/domain/value-object/y-position.ts
export class YPosition {
    private static readonly ERROR_Y_POSITION_INVALID = 'Y position must be a valid number (not NaN)';

    private readonly yPosition: number;

    constructor(yPosition: number) {
        YPosition.validate(yPosition);
        this.yPosition = yPosition;
    }

    private static validate(yPosition: number): void {
        if (Number.isNaN(yPosition)) {
            throw new Error(YPosition.ERROR_Y_POSITION_INVALID);
        }
    }

    public getYPosition(): number {
        return this.yPosition;
    }

    public setYPosition(yPosition: number): YPosition {
        return new YPosition(yPosition);
    }

    public toString(): string {
        return this.yPosition.toString();
    }

    public equals(other: unknown): boolean {
        return (
            other instanceof YPosition &&
            this.yPosition === other.yPosition
        );
    }

    public hashCode(): number {
        const buffer = new ArrayBuffer(8);
        new Float64Array(buffer)[0] = this.yPosition;
        return new Int32Array(buffer)[0];
    }
}

File: back-end/src/main/java/es/ull/project/domain/valueobject/YPosition.java
package es.ull.project.domain.valueobject;

import java.math.BigDecimal;
import java.util.Objects;

/*
 * @brief Represents the Y position of a node in a workflow.
 */
public class YPosition {

    private static final String ERROR_Y_POSITION_INVALID = "Y position must be a valid non-null BigDecimal";
    private final BigDecimal position;

    public YPosition(BigDecimal yPosition) {
        validate(yPosition);
        this.position = yPosition;
    }

    private static void validate(BigDecimal yPosition) {
        if (yPosition == null) {
            throw new IllegalArgumentException(ERROR_Y_POSITION_INVALID);
        }
    }

    public BigDecimal getYPosition() {
        return this.position;
    }

    public YPosition setYPosition(BigDecimal yPosition) {
        return new YPosition(yPosition);
    }

    @Override
    public String toString() {
        return position.toString();
    }

    @Override
    public boolean equals(Object otherObject) {
        if (this == otherObject) {
            return true;
        }
        if (!(otherObject instanceof YPosition)) {
            return false;
        }
        YPosition that = (YPosition) otherObject;
        return this.position.compareTo(that.position) == 0;
    }

    @Override
    public int hashCode() {
        return Objects.hash(position);
    }
}
